main/wallet-backend.js

```diff
@@ -69,119 +83,160 @@ class WalletBackend {
+      console.log('sendTransaction()', {
+        to,
+        amount: amount.toString(),
+        amountMinusGas: amountMinusGas.toString()
+      })
+      return await this.signer.sendTransaction({
+        to,
+        value: amountMinusGas
+      })
     })
-    const messageWithGas = await this.provider.gasEstimateMessageGas(
-      message.toLotusType()
-    )
-    const feeCapStr = messageWithGas.gasFeeCap.toFixed(0, BigNumber.ROUND_CEIL)
-    const feeCap = new FilecoinNumber(feeCapStr, 'attofil')
-    const gas = feeCap.times(messageWithGas.gasLimit)
-    return gas
   }
 
   /**
-   * @param {string} from
    * @param {string} to
-   * @param {FilecoinNumber} amount
+   * @param {ethers.BigNumber} amount
+   * @param {function} fn
    * @returns {Promise<string>}
    */
-  async transferFunds (from, to, amount) {
-    assert(this.provider)
+  async runTransaction (to, amount, fn) {
+    assert(this.signer)
 
     /** @type {FILTransactionProcessing} */
     const transaction = {
       timestamp: new Date().getTime(),
       status: 'processing',
       outgoing: true,
-      amount: amount.toString(),
-      address: to
+      amount: ethers.utils.formatUnits(amount, 18),
+      address: to.startsWith('f1')
+        ? to
+        : delegatedFromEthAddress(to, CoinType.MAIN)
     }
     this.transactions.push(transaction)
     this.onTransactionUpdate()
 
     try {
-      const gasLimit = await this.getGasLimit(from, to, amount)
-      // Ensure transaction has enough gas to succeed
-      const gasOffset = new FilecoinNumber('100', 'attofil')
-      const amountMinusGas = amount.minus(gasLimit).minus(gasOffset)
-      const message = new Message({
-        to,
-        from,
-        nonce: await this.provider.getNonce(from),
-        value: amountMinusGas.toAttoFil(),
-        method: 0,
-        params: ''
-      })
-      const messageWithGas = await this.provider.gasEstimateMessageGas(
-        message.toLotusType()
-      )
-      const lotusMessage = messageWithGas.toLotusType()
-      const msgValid = await this.provider.simulateMessage(lotusMessage)
-      assert(msgValid, 'Message is invalid')
-      const signedMessage = await this.provider.wallet.sign(from, lotusMessage)
-      const { '/': cid } = await this.provider.sendMessage(signedMessage)
-
+      const tx = await fn()
+      console.log({ tx })
+      assert(tx.hash, 'Transaction hash not found')
+      const cid = await this.convertEthTxHashToCid(tx.hash)
+      console.log({ cid })
       transaction.hash = cid
       this.onTransactionUpdate()
-
       return cid
     } catch (err) {
       console.error(err)
       transaction.status = 'failed'
       this.onTransactionUpdate()
-
       throw err
     }
   }
 
+  /**
+   * @param {string} hash
+   * @returns {Promise<string>}
+   */
+  async convertEthTxHashToCid (hash) {
+    console.log('convertEthTxHashToCid', { hash })
+    for (let i = 0; i < 10; i++) {
```

Comment:
Let's use a more descriptive variable than `i` please, to make it more clear what is the intention of this loop. How about `attempts`?

Reply:
good idea, used `attempt`